plugin-lifecycle/src/main/java/com/mapbox/maps/plugin/lifecycle/ViewLifecycleOwner.kt

```diff
@@ -0,0 +1,95 @@
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+import androidx.lifecycle.ViewTreeLifecycleOwner
+import java.lang.ref.WeakReference
+import java.util.concurrent.Flow
+
+/**
+ * A [LifecycleOwner] that merges the hosting lifecycle owner's [Lifecycle] (of an Activity or Fragment)
+ * with the [View]'s drawing surface lifecycle.
+ *
+ * This lifecycle is useful for continuously running tasks that obtain the data and draw it in the [View],
+ * for example by collecting [Flow]s and drawing results in the nested views or on the map.
+ *
+ * In general, this lifecycle reflects the hosting lifecycle owner's [Lifecycle] whenever the `View` is attached
+ * but whenever the `View` is detached and we want to stop the coroutines tied to the lifecycle's scope,
+ * this lifecycle will transition to [Lifecycle.State.CREATED] (and invoke [Lifecycle.Event.ON_STOP])
+ * even if the hosting Activity or Fragment is still running.
+ *
+ * This lifecycle reaches [Lifecycle.State.DESTROYED] only when the hosting lifecycle owner is destroyed.
+ */
+@UiThread
+internal class ViewLifecycleOwner(
+  view: View,
+) : LifecycleOwner {
+  private val viewWeakReference = WeakReference(view)
+  @VisibleForTesting(otherwise = PRIVATE)
+  internal val viewLifecycleRegistry = LifecycleRegistry(this)
+
+  override fun getLifecycle() = viewLifecycleRegistry
+
+  private var isAttached = false
+  private var hostingLifecycleOwner: LifecycleOwner? = null
+
+  private val hostingLifecycleObserver = LifecycleEventObserver { _, event ->
+    val isAtLeastCreated = viewLifecycleRegistry.currentState.isAtLeast(Lifecycle.State.CREATED)
+    if (isAttached || (isAtLeastCreated && event == Lifecycle.Event.ON_DESTROY)) {
+      viewLifecycleRegistry.handleLifecycleEvent(event)
+    }
+  }
+
+  private val attachStateChangeListener = object : View.OnAttachStateChangeListener {
+    override fun onViewAttachedToWindow(p0: View) {
+      doOnAttached(p0)
+    }
+
+    override fun onViewDetachedFromWindow(p0: View) {
+      doOnDetached()
+    }
+  }
+
+  init {
+    view.addOnAttachStateChangeListener(attachStateChangeListener)
+    if (view.isAttachedToWindow) {
+      doOnAttached(view)
+    }
+  }
+
+  private fun doOnAttached(view: View) {
+    if (isAttached) return
+
+    hostingLifecycleOwner?.lifecycle?.removeObserver(hostingLifecycleObserver)
+
+    val hostingLifecycleOwner = ViewTreeLifecycleOwner.get(view)
+      ?: throw IllegalStateException(
+        "Please ensure that the hosting activity/fragment is a valid LifecycleOwner"
+      )
+    viewLifecycleRegistry.currentState = hostingLifecycleOwner.lifecycle.currentState
+    hostingLifecycleOwner.lifecycle.addObserver(hostingLifecycleObserver)
+    this@ViewLifecycleOwner.hostingLifecycleOwner = hostingLifecycleOwner
+    isAttached = true
+  }
+
+  private fun doOnDetached() {
+    if (!isAttached) return
```

Comment:
do we need some sort of warning here before return?

Reply:
I don't think we need to warn anything, these are valid use cases, and we just need to handle them properly.